### ISSUE TITLE
Wildcards in Phishing whitelist and blacklists are ignored

### ISSUE THREAD

**Issue by shawniverson** [MailScanner/v5#184]
Related to https://github.com/MailScanner/v5/issues/108

[EDIT] Wildcards in Phishing blacklists and whitelists are being ignored.

Relevant code snippets:

In Messages.pm:
```
sub DisarmEndtagCallback {

... bunch of ugly regex stuff ...

# (Inserted  for debugging...see what all the above regex result in for $linkurl) on line 7705
MailScanner::Log::InfoLog('linkurl=%s',$linkurl);
```

The above reveals that the $linkurl is in the format of "host.subdomain.example.org" or "host.example.org" or "subdomain.example.org"...

Continuing...

```
sub InPhishingWhitelist {
  my($linkurl) = @_;

  # Quick lookup
  return 1 if $MailScanner::Config::PhishingWhitelist{$linkurl};

  # Trim host. off the front of the hostname
  #while ($linkurl ne "" && $linkurl =~ s/^[^.]+\.//) {
    # And replace it with *. then look it up
    #print STDERR "Looking up *.$linkurl\n";
  #  return 1 if $MailScanner::Config::PhishingWhitelist{'*.' . $linkurl};
  #}

  return 0;
}

MailScanner::Config::PhishingWhitelist is called using $linkurl...

Interesting code commented out above...more on that to follow...

``
# Return 1 if the hostname in $linkurl is in the bad sites file.
sub InPhishingBlacklist {
  my($linkurl) = @_;

  # Quick lookup
  return 1 if $MailScanner::Config::PhishingBlacklist{$linkurl};

  # Trim host. off the front of the hostname
  #while ($linkurl ne "" && $linkurl =~ s/^[^.]+\.//) {
    # And replace it with *. then look it up
    #print STDERR "Looking up *.$linkurl\n";
  #  return 1 if $MailScanner::Config::PhishingBlacklist{'*.' . $linkurl};
  #}

  return 0;
```

MailScanner::Config::PhishingBlacklist is called using $linkurl...

Interesting code commented out above...more on that to follow...

Config.php
```
  # Read in the Phishing Net whitelist.
  # This lists all the hostnames of places to ignore when phishing.
  %PhishingWhitelist = ReadPhishingWhitelist(Value('phishingwhitelist'));
  %PhishingBlacklist = ReadPhishingBlacklist(Value('phishingblacklist'));
...
# In the code for this, it's a direct copy of the ReadPhishingWhitelist() sub,
# so the white and black names are reversed from what would seem logical.
sub ReadPhishingBlacklist {
  my($filename) = @_;

  my($fname, $fh, %whitelist, @blacklist, $counter);

  %whitelist = ();

  # Skip this if they have findphishing = no
  return if MailScanner::Config::IsSimpleValue('findphishing') &&
            !MailScanner::Config::Value('findphishing');

  $filename =~ s/^\s*//g;
  $filename =~ s/\s*$//g;
  return () unless $filename;

  $counter = 0;
  foreach $fname (split(" ", $filename)) {
    next unless $fname;
    $fh = new FileHandle;
    unless (open($fh, "<$fname")) {
      MailScanner::Log::WarnLog("Could not read phishing blacklist file %s", $fname);
      next;
    }

    while(<$fh>) {
      chomp;
      s/^#.*$//;   # Remove comments
      s/^\s*//g;   # Remove leading white space
      s/\s*$//g;   # Remove trailing white space
      s/\s+.*$//g; # Leave only the 1st word
      next if /^$/;
      # Entries in the list starting with "REMOVE " in capitals cause the entry
      # to be forcibly removed from the phishing whitelist.
      if (/^REMOVE\s+(\S+)/i) {
        delete $whitelist{$1};
        push @blacklist, $1;
      } else {
        $whitelist{$_} = 1;
        $counter++;
      }
    }

    # Now process the blacklist
    foreach (@blacklist) {
      delete $whitelist{$_};
    }

    close $fh;
  }
  MailScanner::Log::InfoLog("Read %d hostnames from the phishing blacklists",
                            $counter);

  return %whitelist;
}


# Read the list of hostnames to be ignored when doing phishing tests.
# Pass in the filename. Return the hash.
sub ReadPhishingWhitelist {
  my($filename) = @_;

  my($fname, $fh, %whitelist, @blacklist, $counter);

  %whitelist = ();

  # Skip this if they have findphishing = no
  return if MailScanner::Config::IsSimpleValue('findphishing') &&
            !MailScanner::Config::Value('findphishing');

  $filename =~ s/^\s*//g;
  $filename =~ s/\s*$//g;
  return () unless $filename;

  $counter = 0;
  foreach $fname (split(" ", $filename)) {
    $fh = new FileHandle;
    unless (open($fh, "<$fname")) {
      MailScanner::Log::WarnLog("Could not read phishing whitelist file %s", $fname);
      next;
    }

    while(<$fh>) {
      chomp;
      s/^#.*$//;   # Remove comments
      s/^\s*//g;   # Remove leading white space
      s/\s*$//g;   # Remove trailing white space
      s/\s+.*$//g; # Leave only the 1st word
      next if /^$/;
      # Entries in the list starting with "REMOVE " in capitals cause the entry
      # to be forcibly removed from the phishing whitelist.
      if (/^REMOVE\s+(\S+)/i) {
        delete $whitelist{$1};
        push @blacklist, $1;
      } else {
        $whitelist{$_} = 1;
        $counter++;
      }
    }
  
    # Now process the blacklist
    foreach (@blacklist) {
      delete $whitelist{$_};
    }

    close $fh;
  }
  MailScanner::Log::InfoLog("Read %d hostnames from the phishing whitelist",
                            $counter);

  return %whitelist;
```

Ok, nothing fancy here... reading in the lists pretty much as-is....

Here's the problem:

$linkurl is in the format of "host.subdomain.example.org" or "host.example.org" or "subdomain.example.org"

the phishing.safe.sites.conf are in the following format:

"*.example.org"

But the following code has been commented..
```
  # Trim host. off the front of the hostname
  #while ($linkurl ne "" && $linkurl =~ s/^[^.]+\.//) {
    # And replace it with *. then look it up
    #print STDERR "Looking up *.$linkurl\n";
  #  return 1 if $MailScanner::Config::PhishingWhitelist{'*.' . $linkurl};
  #}
...
  # Trim host. off the front of the hostname
  #while ($linkurl ne "" && $linkurl =~ s/^[^.]+\.//) {
    # And replace it with *. then look it up
    #print STDERR "Looking up *.$linkurl\n";
  #  return 1 if $MailScanner::Config::PhishingBlacklist{'*.' . $linkurl};
  #}
```

and replaced with a simple lookup that does not take into account the *. and the hostname and/or subdomain

This appears to effectively break the lookups to the phishing blacklists and whitelists.

Running more tests..

**Comment by shawniverson** [comment]
Confirmed.  Restoring commented code results in proper function.

Proceeding to regression testing...

**Comment by shawniverson** [comment]
Regression found, backing up a bit.  

There's a rationale as to why the code was changed.  It appears that "Use Stricter Phishing Net" expects full domains to be present in the phishing whitelists and blacklists.  *. entries are purposfully ignored.  in fact, the blacklists are full domains while the whitelists are *.

Looks like restoring this code, although it start functioning, renders "Use Stricter Phishing Net" pointless.

Testing some more....

**Comment by shawniverson** [comment]
Ok, so the root problem is that wildcards are ignored when Use Stricter Phishing Net is enabled.

Testing for expected behavior when Use Stricter Phishing Net is disabled...

**Comment by shawniverson** [comment]
The behavior is also unexpected when the Use Stricter Phishing Net is disabled.  Wildcards are just plain ignored.

**Comment by shawniverson** [comment]
Fixed in commit 7c121ba4934135e5ad4d4518aaf0e2e041dabee5
